### PR TITLE
Add BDD testcases

### PR DESCRIPTION
**Pull request by epapbak**
# Description

- Add BDD test coverage for the main functionalities of the notification service. The added test suites cover:
    - expected behaviour for the CLI arguments
    - ability to display / cleanup records from the DB on demand or when the service starts.
    - basic behaviour of the notification service

- Fix bug in reports retrieval for each cluster in the cluster list 

## Type of change

- Bug fix (non-breaking change which fixes an issue)
- Behavioral tests

## Testing steps

For testing these changes, the following are required:
- local instance of `insights-content-service` listening on port 9092
- kafka broker with the `platform.notifications.ingress` topic created
- prometheus push gateway listening on port 9091
- from `bdd_tests` directory, run:
    -  `NOVENV=1 ./run_tests.sh` if virtual env is not activated
    -   `./run_tests.sh` if virtual env with needed dependencies is activated

## Checklist
* [x] `make before_commit` passes
* [ ] updated documentation wherever necessary
* [x] added or modified tests if necessary
* [ ] updated schemas and validators in [insights-data-schemas](https://github.com/RedHatInsights/insights-data-schemas) in case of input/output change

### CODE REVIEW

**Comment by tisnik** [comment]
Whole venv? :)

**Review by tisnik — COMMENTED** [review]
it's mix from notification writer service BDD, right?

**Comment by epapbak** [comment]
> Whole venv? :)

I have this bad habit: "git add *" :P

**Comment by epapbak** [comment]
> it's mix from notification writer service BDD, right?

Yep. I ended up copying the folder and adding what makes sense, that way there's more coverage than with the 5 scenarios I thought of some months ago.

File: bdd_tests/features/steps/database.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+@when(u"I connect to database named {database} as user {user} with password {password}")
+def connect_to_database(context, database, user, password):
+    """Perform connection to selected database."""
+    connection_string = "dbname={} user={} password={}".format(database, user, password)
+    context.connection = psycopg2.connect(connection_string)
+
+
+@then(u"I should be able to connect to such database")
+def check_connection(context):
+    """Chck the connection to database."""
```
Comment (dpensi) [inline bdd_tests/features/steps/database.py:49]: typo :)

File: bdd_tests/features/steps/database.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+        cursor.execute("SELECT count(*) as cnt from {}".format(table))
+        results = cursor.fetchone()
+        assert len(results) == 1, "Wrong number of records returned: {}".format(len(results))
+        context.query_count = results[0]
+    except Exception as e:
+        raise e
+
+
+@then(u"I should get {expected_count:d} row")
+@then(u"I should get {expected_count:d} rows")
```
Comment (dpensi) [inline bdd_tests/features/steps/database.py:129]: maybe we can use regex https://behave.readthedocs.io/en/stable/regular_expressions.html (I'm not sure and it's not important)

**Review by dpensi — APPROVED** [review]

File: bdd_tests/features/steps/database.py (hunk 2 above)
Comment (epapbak) [reply bdd_tests/features/steps/database.py:129]: It's too much work =)

**Comment by tisnik** [comment]
> > it's mix from notification writer service BDD, right?
> 
> Yep. I ended up copying the folder and adding/removing/modifying what makes sense, that way there's more coverage than with the 5 scenarios I thought of some months ago.

hmm ok. I'm just thinking about some unification/refactoring.

File: bdd_tests/features/steps/common.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+from shutil import which
+import psycopg2
+from psycopg2.errors import UndefinedTable
+
+
+from behave import given, then, when
+
+
+@given(u"the system is in default state")
+def system_in_default_state(context):
```
Comment (juandspy) [inline bdd_tests/features/steps/common.py:24]: Is this function's logic expected? It does nothing

File: bdd_tests/features/steps/common.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    """Check the default system state."""
+    pass
+
+
+@when(u"I look for executable file {filename}")
+def look_for_executable_file(context, filename):
```
Comment (juandspy) [inline bdd_tests/features/steps/common.py:30]: This function actually looks for the executable in PATH, not whether it exists or not. I would change the logic with `file_was_found` function or use `os.path.isfile('file')`.

File: bdd_tests/features/steps/database.py [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+
+
+@then(u"I should be disconnected")
+def check_disconnection(context):
+    """Check that the connection has been closed."""
+    assert context.connection is None, "connection should be closed"
+
+
+@given(u"CCX Notification database is created for user {user} with password {password}")
+@when(u"CCX Notification database is created for user {user} with password {password}")
```
Comment (juandspy) [inline bdd_tests/features/steps/database.py:67]: `When CCX Notification database is created for user {user} with password {password}` is never used in BDD tests.

File: bdd_tests/features/steps/dependencies.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+
+@when(u"I retrieve data from insights-content-service on {host:w}:{port:d}")
+@given(u"insights-content service is available on {host:w}:{port:d}")
+def check_content_service_availability(context, host, port):
+    """Check if insights-content-service is available at given address."""
+
+    if not str(host).startswith("http"):
+        host = "http://" + host
+    x = requests.get(f'{host}:{port}//api/v1/openapi.json')
```
Comment (juandspy) [inline bdd_tests/features/steps/dependencies.py:28]: `f'{host}:{port}//api/v1/openapi.json'` has a `//` typo?

**Review by juandspy — APPROVED** [review]
I cannot check the behavioral tests, but the syntax, structure etc. look good :)

File: bdd_tests/features/steps/common.py (hunk 3 above)
Comment (epapbak) [reply bdd_tests/features/steps/common.py:24]: It's helpful in scenarios where we don't need a specific setup. With this, we can start them with "Given the system is in default state"

File: bdd_tests/features/steps/database.py (hunk 1 above)
Comment (epapbak) [reply bdd_tests/features/steps/database.py:49]: ty

File: bdd_tests/features/steps/common.py (hunk 3 above)
Comment (juandspy) [reply bdd_tests/features/steps/common.py:24]: oh ok, it's like a placeholder

File: bdd_tests/features/steps/common.py (hunk 4 above)
Comment (epapbak) [reply bdd_tests/features/steps/common.py:30]: I'd rather change the name =)

But like Pavel mentioned, this part is copy-pasted from the ccx-notification-writer BDD testcases. I'll leave it as is for now, and take it into account when we do some unification work.

File: bdd_tests/features/steps/database.py (hunk 5 above)
Comment (epapbak) [reply bdd_tests/features/steps/database.py:67]: ty!

File: bdd_tests/features/steps/dependencies.py (hunk 6 above)
Comment (epapbak) [reply bdd_tests/features/steps/dependencies.py]: fixed, ty!

**Review by juandspy — APPROVED** [review]

**Comment by epapbak** [comment]
Merging the change.
Changes for using it in `make test` and in travis CI in follow-up PR

**Pull request merged**